DA/Internal/Desugar.hs

```diff
@@ -97,4 +101,16 @@ class HasToAnyChoice t c r | t c -> r where
 class HasFromAnyChoice t c r | t c -> r where
   _fromAnyChoice : proxy t -> Any -> Optional c
 
-data Implements t i = Implements
+class Implements t i where
+  toInterface : t -> i
+  fromInterface : i -> Optional t
+  toInterfaceContractId : ContractId t -> ContractId i
+  fromInterfaceContractId : ContractId i -> Update (Optional (ContractId t))
```

Comment:
I'm not 100% sure, but I think both `toInterfaceContractId` and `fromInterfaceContractId` don't need to be in the `Implements` typeclass, they can be defined in pure daml (using `Implements t i` as a constraint).

Reply:
that would make sense, though earlier when I tried adding `GHC.Types.primitive @"<primitive>"` as default implementation for all the methods, I noticed that `E{From,To}Interface` needed to be used with concrete types, so it's probably the same for the rest

```
  toInterface = GHC.Types.primitive @"EToInterface"
  fromInterface = GHC.Types.primitive @"EFromInterface"
```

Reply:
`EToInterfaceContractId` and `EFromInterfaceContractId` are fake primitives actually! They're a bunch of generated code that was easier to implement in LFConversion, instead of piping it through ghc. But now that we have a single "Implements" typeclass, I think we can move them out of the compiler and into the stdlib, with the right constraints (fromInterfaceContractId needs a `HasFetch` constraint for example).

Reply:
(In future PRs)

Reply:
aha! That's very interesting, I just saw how they're implemented. I opened https://github.com/digital-asset/daml/issues/11726 so I don't forget about it.